provider/datagen/src/lib.rs

```diff
@@ -246,18 +235,21 @@ pub fn keys_from_bin<P: AsRef<Path>>(path: P) -> std::io::Result<Vec<DataKey>> {
         } else if file[i..].starts_with(icu_provider::trailing_tag!().as_bytes())
             && last_start.is_some()
         {
-            candidates.insert(&file[last_start.unwrap()..i]);
+            if let Some(key) = std::str::from_utf8(&file[last_start.unwrap()..i])
+                .ok()
+                .and_then(key)
+            {
+                result.push(key);
+            }
             i += icu_provider::trailing_tag!().len();
             last_start = None;
         } else {
             i += 1;
         }
     }
-
-    Ok(all_keys()
-        .into_iter()
-        .filter(|k| candidates.contains(k.path().as_bytes()))
-        .collect())
+    // Stability across binary changes.
+    result.sort();
```

Comment:
Nit: You're not checking for uniqueness any more; hopefully there won't be duplicates, but to be sure, can you `.dedup()` the vector after sorting?

Reply:
Sure. We should actually dedupe the input to `datagen`